crates/secrets-core/src/capture.rs

```diff
@@ -0,0 +1,313 @@
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache License, Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2024 Datadog, Inc.
+
+use crate::common::ByteSpan;
+use bstr::{BStr, ByteSlice};
+use std::borrow::Cow;
+use std::iter::{Enumerate, FusedIterator};
+use std::slice::Iter;
+use std::sync::Arc;
+
+/// A list of potentially named captures.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct Captures<'b> {
+    name_slots: Option<Arc<Vec<Option<String>>>>,
+    capture_slots: CaptureSlots<'b>,
+}
+
+impl<'b> Captures<'b> {
+    /// Creates a new [`Captures`] from a `name_slots` and `capture_slots`.
+    pub fn new(
+        name_slots: Option<Arc<Vec<Option<String>>>>,
+        capture_slots: CaptureSlots<'b>,
+    ) -> Captures<'b> {
+        // Captures is often created in a hot-loop, so we can get away with only checking this
+        // invariant in debug mode.
+        debug_assert!(name_slots
+            .as_ref()
+            .map(|name_slots| name_slots.len() == capture_slots.len())
+            .unwrap_or(true));
+        Self {
+            name_slots,
+            capture_slots,
+        }
+    }
```

Comment:
Just curious, why not using [deriver_builder](https://crates.io/crates/derive_builder) for this?